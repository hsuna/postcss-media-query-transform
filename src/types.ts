import type { Input, PluginCreator } from "postcss";

export interface MediaQueryScale {
  query: string;
  scale: number | ((input: Input) => number);
}

export interface UserDefinedOptions {
  mediaQuery?: MediaQueryScale | MediaQueryScale[];
  unitPrecision?: number;
  selectorBlackList?: (string | RegExp)[];
  propList?: (string | RegExp)[];
  transformUnit?: "px" | "rpx" | "rem";
  insert?: "before" | "after";
  exclude?: (string | RegExp)[] | ((filePath: string) => boolean);
  disabled?: boolean;
}

export type PostcssMediaQueryTransform = PluginCreator<UserDefinedOptions>;
