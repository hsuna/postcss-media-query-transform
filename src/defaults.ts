import type { UserDefinedOptions } from "./types";

export const defaultOptions: Required<UserDefinedOptions> = {
  mediaQuery: { query: "(min-width: 400px)", scale: 0.5 },
  unitPrecision: 5,
  selectorBlackList: [],
  propList: ["font", "font-size", "line-height", "letter-spacing"],
  transformUnit: "px",
  exclude: [/node_modules/i],
  disabled: false,
};
