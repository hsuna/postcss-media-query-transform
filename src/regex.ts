// excluding regex trick: http://www.rexegg.com/regex-best-trick.html
export const createUnitRegex = (unit: string) =>
  new RegExp(
    `"[^"]+"|'[^']+'|url\\([^)]+\\)|var\\([^)]+\\)|(\\d+(?:\\.\\d+)?|\\.\\d+)${unit}`,
    "g"
  );
