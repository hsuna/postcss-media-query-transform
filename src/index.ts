import { atRule, type AtRule, Rule } from "postcss";
import { createUnitRegex } from "./regex";
import {
  blacklistedSelector,
  createExcludeMatcher,
  createPropListMatcher,
  createUnitReplace,
  declarationExists,
  getConfig,
  postcssPlugin,
} from "./shared";
import type { PostcssMediaQueryTransform, UserDefinedOptions } from "./types";

export * from "./types";
const plugin: PostcssMediaQueryTransform = (
  options: UserDefinedOptions = {}
) => {
  const {
    mediaQuery,
    unitPrecision,
    transformUnit,
    propList,
    selectorBlackList,
    exclude,
    disabled,
  } = getConfig(options);
  if (disabled || !mediaQuery) {
    return {
      postcssPlugin,
    };
  }

  const satisfyPropList = createPropListMatcher(propList);
  const excludeFn = createExcludeMatcher(exclude);

  return {
    postcssPlugin,
    Once(css) {
      const source = css.source;
      const input = source!.input;
      const filePath = input.file as string;
      const isExcludeFile = excludeFn(filePath);
      if (isExcludeFile) {
        return;
      }

      const mediaQueryList = Array.isArray(mediaQuery)
        ? mediaQuery
        : [mediaQuery];

      const mediaQueryRules = mediaQueryList.map((mediaQuery) => {
        return {
          rule: atRule({
            name: "media",
            params: mediaQuery.query,
          }),
          pxReplace: createUnitReplace(
            typeof mediaQuery.scale === "function"
              ? mediaQuery.scale(input)
              : mediaQuery.scale,
            unitPrecision,
            transformUnit
          ),
        };
      });

      css.walkRules((rule) => {
        if (
          rule.parent?.type === "atrule" &&
          (rule.parent as AtRule).name === "media"
        ) {
          // 如果规则是媒体查询中的规则，则跳过
          return;
        }

        mediaQueryRules.map(({ rule: mediaQueryRule, pxReplace }) => {
          // 克隆规则
          const clonedRule = rule.clone();

          // 遍历克隆规则中的所有声明
          clonedRule.walkDecls((decl) => {
            const rule = decl.parent as Rule;
            if (
              !decl.value.includes(transformUnit) ||
              !satisfyPropList(decl.prop) ||
              blacklistedSelector(selectorBlackList, rule.selector)
            ) {
              // 如果不在规则内，直接移除这条声明
              decl.remove();
              return;
            }

            const value = decl.value.replace(
              createUnitRegex(transformUnit),
              pxReplace
            );

            if (declarationExists(rule, decl.prop, value)) {
              // 如果不在规则内，直接移除这条声明
              decl.remove();
              return;
            }

            // 如果调整后的值与原始值不同，则更新值
            if (value !== decl.value) {
              decl.value = value;
            } else {
              // 如果调整后的值与原始值相同，则移除这条声明
              decl.remove();
            }
          });

          // 如果克隆规则中还有剩余的声明，则将其添加到媒体查询中
          if (clonedRule.nodes?.length) {
            mediaQueryRule.append(clonedRule);
          }
        });
      });

      mediaQueryRules.map(({ rule: mediaQueryRule }) => {
        // 如果媒体查询中有内容，则将其添加到根节点中
        if (mediaQueryRule.nodes?.length) {
          css.append(mediaQueryRule);
        }
      });
    },
  };
};

plugin.postcss = true;

export default plugin;
