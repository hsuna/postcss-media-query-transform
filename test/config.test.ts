import { getConfig } from "src/shared";

describe("config", () => {
  it("getConfig default", () => {
    expect(getConfig()).toMatchSnapshot();
  });

  it("getConfig with wild prop", () => {
    expect(
      getConfig({
        mediaQuery: [
          { query: "(min-width: 400px)", scale: 0.5 },
          { query: "(max-width: 200px)", scale: 2 },
        ],
        propList: ["*"],
      })
    ).toMatchSnapshot();
  });
});
