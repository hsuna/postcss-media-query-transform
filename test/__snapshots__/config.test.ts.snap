// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`config > getConfig default 1`] = `
{
  "disabled": false,
  "exclude": [
    /node_modules/i,
  ],
  "mediaQuery": {
    "query": "(min-width: 400px)",
    "scale": 0.5,
  },
  "propList": [
    "font",
    "font-size",
    "line-height",
    "letter-spacing",
  ],
  "selectorBlackList": [],
  "transformUnit": "px",
  "unitPrecision": 5,
}
`;

exports[`config > getConfig with wild prop 1`] = `
{
  "disabled": false,
  "exclude": [
    /node_modules/i,
  ],
  "mediaQuery": [
    {
      "query": "(min-width: 400px)",
      "scale": 0.5,
    },
    {
      "query": "(max-width: 200px)",
      "scale": 2,
    },
  ],
  "propList": [
    "*",
  ],
  "selectorBlackList": [],
  "transformUnit": "px",
  "unitPrecision": 5,
}
`;
